test/realm_test.dart

```diff
@@ -1567,6 +1566,75 @@ Future<void> main([List<String>? args]) async {
     //test the realm can be opened.
     final realm = getRealm(Configuration.disconnectedSync([Product.schema], path: path, encryptionKey: key));
   });
+
+  test('Realm.refresh', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    final result = realm.refresh();
+    expect(result, true);
+  });
+
+  test('Realm.onRefresh sync transaction', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    var called = false;
+    final transaction = realm.beginWrite();
+    realm.refreshAsync().then((_) => called = true);
+    realm.add(Person("name"));
+    transaction.commit();
+
+    await Future<void>.delayed(Duration(milliseconds: 1));
+    expect(called, true, reason: "beginWrite onRefresh failed");
+
+    called = false;
+    realm.write(() {
+      realm.refreshAsync().then((_) => called = true);
+      realm.add(Person("name"));
+    });
+
+    await Future<void>.delayed(Duration(milliseconds: 1));
+    expect(called, true, reason: "write onRefresh failed");
+  });
+
+  test('Realm.onRefresh async transaction', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    bool called = false;
+    final transaction = await realm.beginWriteAsync();
+    realm.refreshAsync().then((_) => called = true);
+    realm.add(Person("name"));
+    await transaction.commitAsync();
+    expect(called, true);
+  });
+
+   test('Realm.onRefresh is called if registered outside a transaction', () async {
+    final realm = getRealm(Configuration.local([Person.schema]));
+    bool called = false;
+
+    realm.refreshAsync().then((_) => called = true);
+
+    realm.beginWrite().commit();
+
+    await Future<void>.delayed(Duration(milliseconds: 1));
+
+    expect(called, true);
+
+
+    realm.refreshAsync().then((_) => called = true);
+
+    realm.write(() {});
+
+    await Future<void>.delayed(Duration(milliseconds: 1));
+
+    expect(called, true);
```

Comment:
```suggestion
  final realm = getRealm(Configuration.local([Person.schema]));

    final first = expectLater(realm.refreshAsync(), completes);
    realm.beginWrite().commit();
    await (first);

    final second = expectLater(realm.refreshAsync(), completes);
    realm.write(() {});
    await (second);
```

Reply:
This changes the test too much as it is not readable what is happening and what is the expectation. Lets leave it as I have written so it is more clear what goes after the other.

Reply:
I believe it is clearer like I write it. It is also shorter. I would like us to stop using the flag based approach, and spread the knowledge of how to do it right.

Reply:
I'm not sure what this test is supposed to test (in either of the forms). `realm.refreshAsync()` should complete synchronously when we're already at the latest version, which is pretty much always when using a single isolate. So it will have resolved already before we even get to `realm.write`. If we want to verify that it actually does something, we need to spin up a second isolate, open the same realm, write a bunch of data in it, then commit the transaction, signal the first isolate, then assert the data is not visible, call refresh async, assert that the data is visible.

Reply:
this test is the same as what we have in Core. Just using Dart

Reply:
@nirinchev Currently calling `await realm.refreshAsync()` today will block until something is actually written to the realm. So it tests that.. which may well be a bug in the C-api.

Reply:
Yeah, it's not the intended behavior though. There's no obvious situation in which this is going to be useful behavior.

Reply:
I can see value in knowing when the next transaction completes, but I agree that is not what this API is intended for.

Reply:
The test suggested by @nirinchev is added.